GVFS/GVFS.Platform.POSIX/POSIXFileBasedLock.cs

```diff
@@ -33,91 +33,91 @@ public override bool TryAcquireLock()
-
-            public const int LockSh = 1; // #define LOCK_SH   1    /* shared lock */
-            public const int LockEx = 2; // #define LOCK_EX   2    /* exclusive lock */
-            public const int LockNb = 4; // #define LOCK_NB   4    /* don't block when locking */
-            public const int LockUn = 8; // #define LOCK_UN   8    /* unlock */
-
-            public const int InvalidFileDescriptor = -1;
-
+                }
+            }
+
+            if (NativeMethods.FLock(this.lockFileDescriptor, NativeMethods.LockEx | NativeMethods.LockNb) != 0)
+            {
+                int errno = Marshal.GetLastWin32Error();
+                if (errno != NativeMethods.EIntr && errno != NativeMethods.EWouldBlock)
+                {
+                    EventMetadata metadata = this.CreateEventMetadata(errno);
+                    this.Tracer.RelatedWarning(
+                        metadata,
+                        $"{nameof(POSIXFileBasedLock)}.{nameof(this.TryAcquireLock)}: Unexpected error when locking file");
+                }
+
+                return false;
+            }
+
+            return true;
+        }
+
+        public override void Dispose()
+        {
+            if (this.lockFileDescriptor != NativeMethods.InvalidFileDescriptor)
+            {
+                if (NativeMethods.Close(this.lockFileDescriptor) != 0)
+                {
+                    // Failures of close() are logged for diagnostic purposes only.
+                    // It's possible that errors from a previous operation (e.g. write(2))
+                    // are only reported in close().  We should *not* retry the close() if
+                    // it fails since it may cause a re-used file descriptor from another
+                    // thread to be closed.
+
+                    int errno = Marshal.GetLastWin32Error();
+                    EventMetadata metadata = this.CreateEventMetadata(errno);
+                    this.Tracer.RelatedWarning(
+                        metadata,
+                        $"{nameof(POSIXFileBasedLock)}.{nameof(this.Dispose)}: Error when closing lock fd");
+                }
+
+                this.lockFileDescriptor = NativeMethods.InvalidFileDescriptor;
+            }
+        }
+
+        private EventMetadata CreateEventMetadata(int errno = 0)
+        {
+            EventMetadata metadata = new EventMetadata();
+            metadata.Add("Area", nameof(POSIXFileBasedLock));
+            metadata.Add(nameof(this.LockPath), this.LockPath);
+            if (errno != 0)
+            {
+                metadata.Add(nameof(errno), errno);
+            }
+
+            return metadata;
+        }
+
+        private static class NativeMethods
+        {
+            // #define O_WRONLY    0x0001      /* open for writing only */
+            public const int OpenWriteOnly = 0x0001;
+
+            // #define O_CREAT     0x0200      /* create if nonexistant */
+            public const int OpenCreate = 0x0200;
```

Comment:
One problem here is that these constants aren't the same on Linux as on BSD/macOS.

In particular, this should be `OpenCreate = 0x0040` on Linux, and below should have `EWouldBlock = 11`, and use `uint` and `ToUInt32` for the `mode_t` values.  See b02e4a3845 for what I believe are the relevant correct values, but that could use a double-check from someone else, for sure.

As a result, I expect this class should be divvied up into Linux and Mac variants, perhaps as a base common POSIX class (with things like `public const int InvalidFileDescriptor = -1`) and then the platform-specific values defined in separate classes.

Reply:
Super helpful commit link, thank you! It'll be easier to know what to refactor when it does come time to add the Linux platform on top of these commits -- for now I'll mark POSIXFileBasedLock as abstract and add an empty implementation class for Mac so we can shift specific stuff into it when we refactor later.